### PR TITLE
Handle YAML configuration format on configuration file

### PR DESCRIPTION
**Pull request by cosmo0920**
**Which issue(s) this PR fixes**: 
Fixes https://github.com/fluent/fluentd/issues/554
Revised #2818

**What this PR does / why we need it**: 

Recently, fluent-bit supports YAML configuration in https://github.com/fluent/fluent-bit/pull/4621.

Now, it is reasonable to handle YAML configuration format on Fluentd, too.
This is because YAML config format is more kubernetes friendly and machine friendly.
It's time to discuss to support YAML configuration format again.
 
**Docs Changes**:
https://github.com/fluent/fluentd-docs-gitbook/pull/402

**Release Note**: 

Same as title

### CODE REVIEW

File: lib/fluent/config.rb [hunk 1]
```diff
@@ -25,7 +26,18 @@ module Config
     # @param additional_config [String] config which is added to last of config body
     # @param use_v1_config [Bool] config is formatted with v1 or not
     # @return [Fluent::Config]
-    def self.build(config_path:, encoding: 'utf-8', additional_config: nil, use_v1_config: true)
+    def self.build(config_path:, encoding: 'utf-8', additional_config: nil, use_v1_config: true, type: nil, guess_type: nil)
+      if guess_type
+        config_fext = File.extname(config_path.dup)
```
Comment (ashie) [inline lib/fluent/config.rb:31]: I prefer `config_file_ext` or `config_ext`.

Comment (ashie) [reply lib/fluent/config.rb:31]: > `config_path.dup`

`.dup` doesn't seem needed.

**Comment by cosmo0920** [comment]
@ashie @kenhys How's it going to review this PR?

**Comment by ashie** [comment]
I think #3535 and this are featured functions of next release, v1.15.
#3535 has been landed just now, so I'll start to review this from now :smiling_face_with_tear:

**Comment by ashie** [comment]
(Although I already reviewed it roughly and it seems fine, I'll check the details more.)

File: lib/fluent/command/fluentd.rb [hunk 2]
```diff
@@ -126,6 +126,18 @@
...
 
+op.on('--config-file-type VALU', 'file type of fluentd. yaml or yml') { |s|
+  if (s == 'yaml') || (s == 'yml')
+    opts[:config_file_type] = s.to_sym
+  else
+    usage '--config-file-type accepts yaml or yml'
+  end
+}
+
+op.on('--guess-config-file-type', "Guess config file type that is yaml or not (default)", TrueClass) {|b|
```
Comment (ashie) [inline lib/fluent/command/fluentd.rb:137]: How about removing this and add `guess` to `--config-file-type` (and make it default).

Comment (cosmo0920) [reply lib/fluent/command/fluentd.rb:137]: Sure, I'll change them.

Comment (cosmo0920) [reply lib/fluent/command/fluentd.rb]: Fixed in https://github.com/fluent/fluentd/pull/3712/commits/49ec02315311c94ba3550c0b752ad24deaaaee0e

Comment (ashie) [reply lib/fluent/command/fluentd.rb]: :+1:

Comment (ashie) [reply lib/fluent/command/fluentd.rb]: Ah, sorry it's a bit different what I thought.
I intended adding `:guess` (`--config-file-type=guess`), how about this?

Comment (cosmo0920) [reply lib/fluent/command/fluentd.rb]: Ah, I see. I'll try to implement what you thought.

**Review by ashie — CHANGES_REQUESTED** [review]

File: lib/fluent/command/fluentd.rb (hunk 2 above)
Comment (cosmo0920) [reply lib/fluent/command/fluentd.rb]: Really fixed in https://github.com/fluent/fluentd/pull/3712/commits/6574387ef2c78363dc23640df14d9886d5606bba

File: lib/fluent/config/yaml_parser/parser.rb [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    module YamlParser
+      class Parser
+        def initialize(config, indent: 2)
+          @base_indent = indent
+          @config = config
+        end
+
+        def build
+          s = @config['system'] && system_config_build(@config['system'])
+          c = @config['config'] && config_build(@config['config'], root: true)
```
Comment (ashie) [inline lib/fluent/config/yaml_parser/parser.rb:30]: Always requires `config` parent for `label`, `source`, `filter`, `match` and `worker` is bother and doesn't compatible with Fluentd's format. It makes hard to understand how to convert.

Comment (cosmo0920) [reply lib/fluent/config/yaml_parser/parser.rb:30]: This format limitations is coming from _YAML format specification and its parser simplicity_.
Fluentd config format does not have 1-to-1 correspondence for YAML config format for now.
It is quite hard to implement to arbitrary place handling for `label`, `source`, `filter`, `match` and `worker`. directives.

I didn't find no mono morphic way from Fluentd config format to YAML format.

Comment (cosmo0920) [reply lib/fluent/config/yaml_parser/parser.rb:30]: FYI: Fluent Bit uses similar approach to handle data pipeline:

https://github.com/fluent/fluent-bit/pull/4734

Fluent Bit ini like format should use `[INPUT]`, `[OUTOUT]`, `[FILTER]` sections as a top level settings on the original config implementation but its YAML format does not use them as top level objects.

Comment (cosmo0920) [reply lib/fluent/config/yaml_parser/parser.rb:30]: And we already prepared from Fluentd configuration format to YAML format translation documents here:
https://github.com/fluent/fluentd-docs-gitbook/pull/402

Comment (cosmo0920) [reply lib/fluent/config/yaml_parser/parser.rb:30]: Our assumed use-case is:
This format is mainly used from Golang's struct Marshal/Unmarshal not by writing by hands.

Comment (ashie) [reply lib/fluent/config/yaml_parser/parser.rb:30]: I got it. For example, the following is invalid as yaml:
```yml
- label:
    $name: "@FLUENT_LOG"
   - match:
       $type: "stdout"
       $tag: tag1.**
  - match:
      $type: "null"
      $tag: tag2.**
```

**Comment by cosmo0920** [comment]
Any concerns?

**Review by ashie — APPROVED** [review]

**Comment by ashie** [comment]
:+1:

**Comment by cosmo0920** [comment]
Thanks!!

**Comment by daipom** [comment]
https://github.com/fluent/fluentd/issues/4463#issuecomment-2054279009

> it appears that YAML format doesn't support the $log_level

If someone remembers something about this, please let me know.

File: lib/fluent/config/yaml_parser/parser.rb [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+          end
+
+          if (v = config.delete('$label'))
+            sb.add_line('@label', v)
+          end
+
+          if (v = config.delete('$id'))
+            sb.add_line('@id', v)
+          end
+
```
Comment (cosmo0920) [inline lib/fluent/config/yaml_parser/parser.rb:140]: @daipom Probably, we should add handler for `$log_level` here. Do you have a bandwidth for this?

Comment (daipom) [reply lib/fluent/config/yaml_parser/parser.rb:140]: Thanks!

I don't think this is a serious problem.
I will work on other issues for a while.

Comment (cosmo0920) [reply lib/fluent/config/yaml_parser/parser.rb:140]: Sure. Fair enough.

**Pull request merged**